src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs

```diff
@@ -759,5 +761,38 @@ public static DocumentId GetDocumentIdForTree(SyntaxTree tree)
             s_syntaxTreeToIdMap.TryGetValue(tree, out var id);
             return id;
         }
+
+        private static void CheckTree(
+            SyntaxTree newTree,
+            SourceText newText,
+            SyntaxTree oldTree = null,
+            SourceText oldText = null)
+        {
+            // this should be always true
+            if (newTree.Length == newText.Length)
```

Comment:
we fire NFW for the root cause explicitly with relevant info saved in the heap to make repro easier.